File: python/src/simpson_integration.py
def integral_evaluation(N, a, b, func):
    # Check if N is an even integer
    if not isinstance(N, int) or N % 2 != 0:
        raise TypeError('N must be an even integer')
    if not isinstance(a, (int, float)) or not isinstance(b, (int, float)):
        raise TypeError('a and b must be numbers')
    if N <= 0:
        raise ValueError('N has to be >= 2')
    if a > b:
        raise ValueError('a must be less or equal than b')
    if a == b:
        return 0

    # Calculate the step h
    h = (b - a) / N

    # Find interpolation points
    xi = a  # initialize xi = x0
    points_array = []

    # Find the sum {f(x0) + 4*f(x1) + 2*f(x2) + ... + 2*f(xN-2) + 4*f(xN-1) + f(xN)}
    for i in range(N + 1):
        if i == 0 or i == N:
            temp = func(xi)
        elif i % 2 == 0:
            temp = 2 * func(xi)
        else:
            temp = 4 * func(xi)
        points_array.append(temp)
        xi += h

    # Calculate the integral
    result = h / 3 * sum(points_array)

    if result != result:  # Check for NaN
        raise ValueError("Result is NaN. The input interval doesn't belong to the function's domain")

    return result

File: javascript/src/simpson-integration.js
function integralEvaluation(N, a, b, func) {
    let isNEven = true
    if (N % 2 !== 0) isNEven = false

    if (!Number.isInteger(N) || Number.isNaN(a) || Number.isNaN(b)) {
        throw new TypeError('Expected integer N and finite a, b')
    }
    if (!isNEven) {
        throw Error('N is not an even number')
    }
    if (N <= 0) {
        throw Error('N has to be >= 2')
    }

    if (a > b) {
        throw Error('a must be less or equal than b')
    }
    if (a === b) return 0

    const h = (b - a) / N

    let xi = a 
    const pointsArray = []

    let temp
    for (let i = 0; i < N + 1; i++) {
        if (i === 0 || i === N) temp = func(xi)
        else if (i % 2 === 0) temp = 2 * func(xi)
        else temp = 4 * func(xi)

        pointsArray.push(temp)
        xi += h
    }

    let result = h / 3
    temp = pointsArray.reduce((acc, currValue) => acc + currValue, 0)

    result *= temp

    if (Number.isNaN(result)) {
        throw Error(
            "Result is NaN. The input interval doesn't belong to the functions domain"
        )
    }

    return result
}

export { integralEvaluation }